keps/sig-auth/20191101-add-request-id-to-k8s-logs.md

```diff
@@ -0,0 +1,322 @@
+     - [Design overview of Control Request-ID](#design-overview-of-control-request-id)
+     - [Detail design of Control Request-ID](#detail-design-of-control-request-id)
+ - [Test Plan](#test-plan)
+ - [Migration / Graduation Criteria](#migration--graduation-criteria)
+   - [Alpha](#alpha)
+   - [Beta](#beta)
+   - [GA](#ga)
+
+<!-- /toc -->
+
+## Summary
+
+This KEP proposes a new unique logging meta-data into all Kubernetes logs. It makes us
+more easy to identify specific logs related to a single user operation (such as
+`kubectl apply -f <my-pod.yaml>`). This feature is similar to
+[Global request ID](https://docs.openstack.org/api-guide/compute/faults.html) for
+OpenStack. It greatly reduces investigation cost.
+
+## Motivation
+
+### Target User
+
+Support team in k8s Service Provider
+
+### Target User's objective
+
+We'd like to resolve quickly for end users' problem.
+
+Tracking logs among each Kubernetes component related to specific an user operation is very tough work. It is necessary to match logs by basically using timestamps and object's name as hints. If multiple users throw many API requests at the same time, it is very difficult to track logs across each Kubernetes component log. 
+
+It is difficult that support team in k8s Service Provider resolve end user's problem quickly in the above. Therefore, we'd like to add a new identifier which is unique to each user operation. This feature is useful for the following use cases:
+
+#### Case 1
+
+In case of insecure or unauthorized operation happens, it is necessary to
+identify what effect that operation caused. This proposed feature helps identify
+what happened at each component or server by each insecure / unauthorized API
+request. We can collect these logs as an evidence. This is similar to the 
+[Auditing](https://kubernetes.io/docs/tasks/debug-application-cluster/audit/), 
+except for the following points.
+
+ - Audit only collects information about http request sending and receiving in kube-apiserver, so it can't track internal work of each component.
+ - Audit logs can't be associated to logs related to user operation (kubectl operation), because auditID is different for each http request.
+
+
+#### Case 2
+
+If the container is terminated by OOM killer, there is a case to break down the
+issue into parts(Pod or Kubernetes) from the messages related OOM killer on host logs
+and the API processing just before OOM killer. If the cause is that some unknown
+pod creations, it is helpful to detect the root API request and who called this
+request.
+
+### Goals
+
+ - Adding a Request-ID into each K8s component log.
+ - The Request-ID is unique to a kubectl operation.
+   - (One kubectl operation by user causes multiple API requests and klog calls. Request-ID has same value in these klog calls.)
+ - Control enabled/disabled Request-ID feature(Request-ID feature is disabled on default to avoid an impact for existing user).
```

Comment:
Usually the process of moving from:
disabled by default  -> enabled by default -> always enabled
maps to
alpha -> beta -> GA

Unless you are proposing something outside of this, it probably doesn't need to be mentioned in goals.

Reply:
I think you are pointing `feature gate`.
I considered whether `feature gate` can satisfy our objective or not, but I'd like to allow the user to adjust the trade-off with log size. (This is similar to the concept of log level), so I'd like to implement the feature which can control request-id feature.